bundles/org.openhab.binding.e3dc/src/main/java/org/openhab/binding/e3dc/internal/handler/BaseHandler.java

```diff
@@ -0,0 +1,116 @@
+import org.openhab.binding.e3dc.internal.modbus.DataListener;
+import org.openhab.binding.e3dc.internal.modbus.ModbusDataProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link BaseHandler} Basic Handler for all things
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public abstract class BaseHandler extends BaseThingHandler implements DataListener {
+    private final Logger logger = LoggerFactory.getLogger(BaseHandler.class);
+    private @Nullable ThingHandlerCallback thingHandlerCallback;
+    private @Nullable ModbusDataProvider modbusDataProvider;
+
+    // Unit test purposes - check if dataAvailable callback was successful
+    public boolean isUpdated = false;
+    public boolean isInitialized = false;
+
+    public BaseHandler(Thing thing) {
+        super(thing);
+    }
+
+    public void initialize(DataType t) {
+        updateStatus(ThingStatus.UNKNOWN);
+        scheduler.execute(() -> {
+            Bridge b = getBridge();
+            if (b != null) {
+                BridgeHandler bridgeHandler = b.getHandler();
+                if (bridgeHandler != null) {
+                    if (t.equals(DataType.INFO)) {
+                        ModbusDataProvider localModbusDataProvider = ((E3DCDeviceThingHandler) bridgeHandler)
+                                .getInfoDataProvider();
+                        modbusDataProvider = localModbusDataProvider;
+                        localModbusDataProvider.addDataListener(this);
+                        updateStatus(ThingStatus.ONLINE);
+                    } else if (t.equals(DataType.DATA)) {
+                        ModbusDataProvider localModbusDataProvider = ((E3DCDeviceThingHandler) bridgeHandler)
+                                .getDataProvider();
+                        modbusDataProvider = localModbusDataProvider;
+                        localModbusDataProvider.addDataListener(this);
+                        updateStatus(ThingStatus.ONLINE);
+                    }
+                } else {
+                    logger.warn("BridgeHandler not found");
+                }
+            } else {
+                logger.warn("Bridge not found");
+            }
+            isInitialized = true;
+        });
+    }
+
+    @Override
+    public void dispose() {
+        ModbusDataProvider localModbusDataProvider = modbusDataProvider;
+        if (localModbusDataProvider != null) {
+            localModbusDataProvider.removeDataListener(this);
+        }
+    }
+
+    @Override
+    public void setCallback(@Nullable ThingHandlerCallback thingHandlerCallback) {
+        super.setCallback(thingHandlerCallback);
+        this.thingHandlerCallback = thingHandlerCallback;
+        // logger.info("ThingCallbackHandler received {}", thingHandlerCallback);
+    }
+
+    @Override
+    public @Nullable Bridge getBridge() {
+        ThingUID bridgeUID = thing.getBridgeUID();
+        synchronized (this) {
```

Comment:
What's your intention by making this synchronized? The callback is set before `initialize()` is invoked.

Reply:
Obsolete due to Modbus rework